crypto/bls12/bls12.go

```diff
@@ -155,231 +131,257 @@ func (agg AggregateSignature) Bitfield() crypto.Bitfield {
-	sig.participants.ForEach(func(id hotstuff.ID) {
-		replica, ok := bc.mods.Configuration().Replica(id)
-		if !ok {
-			return
-		}
-		pubKeys = append(pubKeys, replica.PublicKey().(*PublicKey))
-	})
-	ps, err := bls12.NewG2().HashToCurve(hash[:], domain)
+	g2 := bls12.NewG2()
+	messagePoint, err := g2.HashToCurve(message, domainTag)
 	if err != nil {
-		bc.mods.Logger().Error(err)
 		return false
 	}
 	engine := bls12.NewEngine()
-	engine.AddPairInv(&bls12.G1One, &sig.sig)
-	for _, pub := range pubKeys {
-		engine.AddPair(pub.p, ps)
-	}
+	engine.AddPairInv(&bls12.G1One, signature)
+	engine.AddPair(pubKey.p, messagePoint)
 	return engine.Result().IsOne()
 }
 
-// TODO: I'm not sure to what extent we are vulnerable to a rogue public key attack here.
-// As far as I can tell, this is not a problem right now because we do not yet support reconfiguration,
-// and all public keys are known by all replicas.
+func (bls *bls12Base) popProve() *bls12.PointG2 {
+	pubKey := bls.getPrivateKey().Public().(*PublicKey)
+	proof, err := bls.coreSign(pubKey.ToBytes(), domainPOP)
+	if err != nil {
+		bls.mods.Logger().Panicf("failed to generate proof of possession: %v", err)
+	}
+	return proof
+}
 
-// VerifyThresholdSignature verifies a threshold signature.
-func (bc *bls12Crypto) VerifyThresholdSignature(signature consensus.ThresholdSignature, hash consensus.Hash) bool {
-	sig, ok := signature.(*AggregateSignature)
+func (bls *bls12Base) popVerify(pubKey *PublicKey, proof *bls12.PointG2) bool {
+	return bls.coreVerify(pubKey, pubKey.ToBytes(), proof, domainPOP)
+}
+
+func (bls *bls12Base) checkPop(replica consensus.Replica) (valid bool) {
+	defer func() {
+		if !valid {
+			bls.mods.Logger().Debugf("Invalid proof-of-possession for replica %d", replica.ID())
+		}
+	}()
+
+	b, ok := replica.Metadata()[popMetadataKey]
 	if !ok {
+		bls.mods.Logger().Debugf("Missing proof-of-possession for replica: %d", replica.ID())
 		return false
 	}
-	pubKeys := make([]*PublicKey, 0)
-	sig.participants.ForEach(func(id hotstuff.ID) {
-		replica, ok := bc.mods.Configuration().Replica(id)
-		if !ok {
-			return
-		}
-		pubKeys = append(pubKeys, replica.PublicKey().(*PublicKey))
-	})
-	ps, err := bls12.NewG2().HashToCurve(hash[:], domain)
+
+	var key strings.Builder
+	key.WriteString(b)
+	_, _ = key.Write(replica.PublicKey().(*PublicKey).ToBytes())
+
+	bls.mut.RLock()
+	valid, ok = bls.popCache[key.String()]
+	bls.mut.RUnlock()
+	if ok {
+		return valid
+	}
+
+	p, err := bls12.NewG2().FromCompressed([]byte(b))
 	if err != nil {
-		bc.mods.Logger().Error(err)
 		return false
 	}
-	if len(pubKeys) < bc.mods.Configuration().QuorumSize() {
+
+	valid = bls.popVerify(replica.PublicKey().(*PublicKey), p)
+
+	bls.mut.Lock()
+	bls.popCache[key.String()] = valid
+	bls.mut.Unlock()
+
+	return valid
+}
+
+func (bls *bls12Base) coreAggregateVerify(publicKeys []*PublicKey, messages [][]byte, signature *bls12.PointG2) bool {
+	n := len(publicKeys)
+	if n != len(messages) && n <= 1 {
```

Comment:
Do you need `&& n <= 1`? Does this method not work for `n=1` and `len(publicKeys) == len(messages)`? Maybe it would help to add a comment to explain the limitations of this method.

Reply:
Oops, this is actually wrong. n >= 1 is a precondition of the function, and the length of the public keys and messages slices must be equal